### ISSUE TITLE
Bug: `no-unused-vars` is triggered for arguments of function type definitions

### ISSUE THREAD

**Issue by iFreilicht** [eslint/eslint#17849]
### Environment

Node version: v20.4.0
npm version: v9.7.2
Local ESLint version: v8.55.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 23.1.0

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
module.exports = {
  root: true,
  env: { browser: true, es2020: true },
  extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended'],
  ignorePatterns: ['dist', '.eslintrc.cjs'],
  parser: '@typescript-eslint/parser',
  rules: {
    'no-unused-vars': ['warn'],
  },
};
```
</details>

```js
export type MyType = {
  exec: (command: string) => Promise<void>;
};
```

### What did you expect to happen?

I would expect this to be fine. `command` is only part of the type definition, it can't be "used" in any way.

### What actually happened?

I get the following warning:

```
npm exec eslint testfile.ts

/Users/feuh/repos/nix-demo/presentation/testfile.ts
  2:10  warning  'command' is defined but never used  no-unused-vars

✖ 1 problem (0 errors, 1 warning)
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/eslint-repro-2023-12-13

### Participation

- [ ] I am willing to submit a pull request for this issue.
